Sniper scope

Badulundabad opened the pull request:
- added sniper_scope.xml;
- added SniperScope class;
- added TORCustomBattleAgentStatCalculateModel to change zoom in custom battle;
- overrided GetMaxCamerZoom in TORAgentStatCalculateModel to change zoom in campaign battle;
- adjusted CustomCrosshairMissionBehavior for sniper scope.

Badulundabad committed:
added test code

Badulundabad committed:
added sniper scope xml

Badulundabad committed:
added new agent stat model, added logic of sniper scope

Badulundabad committed:
added SniperScope class, adjusted crosshair mission behavior

Badulundabad committed:
Merge branch 'development' into 3d_sniper_scope_logic

Badulundabad committed:
dll

Badulundabad committed:
added hitbox to sniper scope

Badulundabad committed:
returned missed line

Z3rca commented on CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs:
```diff
@@ -150,5 +141,37 @@ public override void OnPhotoModeDeactivated()
+        private bool IsPlayerAimingWithSniperRifle()
+        {
+            return Mission.CameraIsFirstPerson &&
+                   Input.IsKeyDown(InputKey.LeftShift) &&
```
note. we really need to try to implement the custom input system. Having hardcoded input checks will be problematic in future

Badulundabad replied:
this method was removed, because it's not needed

Badulundabad replied:
but i understand your point, moreover we have a lot of methods with using native Input. Anyway i think we will come to implementation our own input someday

Badulundabad committed:
sniper_cope_branch. made a lot of changes

CustomCrosshairMissionBehavior:
- renamed and changed IsPlayerAimingWithSniperScope;
- added IsRightAngleToShoot method;
- added IsUsingSniperScope property (+field).

Crosshair:
- adjusted logic of Tick method;
- removed GetShouldCrosshairBeVisible and
IsMissionScreenUsingCustomCamera methods.

Changed GetMaxZoom method in TORCustomBattleAgentStatCalculateModel and TORAgentStatCalculateModel;

Badulundabad committed:
fixed CustomCrosshairMissionBehavior

Badulundabad committed:
removed test code

Badulundabad committed:
Merge branch 'development' into 3d_sniper_scope_logic

hunharibo commented on CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs:
```diff
@@ -150,5 +163,44 @@ public override void OnPhotoModeDeactivated()
+            if (Agent.Main.IsAbilityUser() && (_abilityComponent = Agent.Main.GetComponent<AbilityComponent>()) != null)
+            {
+                _missionLogic = Mission.Current.GetMissionBehavior<AbilityManagerMissionLogic>();
+                _abilityComponent.CurrentAbilityChanged += (crosshair) =>
```
You should remove the event listener in the Finalize method.

Badulundabad replied:
yeah, you're totally right

hunharibo reviewed (approved):
One minor comment, nicely cleaned up otherwise :)

Badulundabad committed:
fixed issue mentioned by hunharibo

Badulundabad committed:
removed excess method

Badulundabad committed:
fixed CustomCrosshairMissionBehavior

Z3rca commented on CSharpSourceCode/Battle/Crosshairs/Crosshair.cs:
```diff
@@ -44,132 +43,119 @@ public void FinalizeCrosshair()
-            if (Mission.Current.Mode != MissionMode.Conversation && Mission.Current.Mode != MissionMode.CutScene && Mission.Current.Mode != MissionMode.Deployment && Mission.Current.MainAgent != null && !_missionScreen.IsViewingCharacter() && !this.IsMissionScreenUsingCustomCamera())
+            _crosshairVM.CrosshairType = BannerlordConfig.CrosshairType;
+            Agent mainAgent = Mission.Current.MainAgent;
+            double num = (double)(_missionScreen.CameraViewAngle * 0.017453292f);
```
please use a const for variables to get rid of magic numbers

Badulundabad replied:
it's native method, with unknown logic

Z3rca commented on CSharpSourceCode/Battle/Crosshairs/Crosshair.cs:
```diff
@@ -44,132 +43,119 @@ public void FinalizeCrosshair()
+            Agent mainAgent = Mission.Current.MainAgent;
+            double num = (double)(_missionScreen.CameraViewAngle * 0.017453292f);
+            double accuracy = 2.0 * Math.Tan((double)(mainAgent.CurrentAimingError + mainAgent.CurrentAimingTurbulance) * (0.5 / Math.Tan(num * 0.5)));
+            _crosshairVM.SetProperties(accuracy, (double)(1f + (_missionScreen.CombatCamera.HorizontalFov - 1.5707964f) / 1.5707964f));
```
same here

Badulundabad replied:
the same)

Z3rca reviewed (commented):
mostly okay, just get rid of the magic numbers and replace them with  some const.

Badulundabad committed:
added finalizing method to SniperScope

Badulundabad committed:
added constants

Badulundabad committed:
Merge branch 'development' into 3d_sniper_scope_logic
